pkg/controllers/mcs/common.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/karmada-io/karmada/pkg/util/fedinformer/keys"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+	"github.com/karmada-io/karmada/pkg/util/names"
+)
+
+func cleanupDerivedWork(c client.Client, key keys.FederatedKey) error {
+	clusters := &clusterv1alpha1.ClusterList{}
+	if err := c.List(context.TODO(), clusters); err != nil {
+		return err
+	}
+	for _, cluster := range clusters.Items {
+		key.Cluster = cluster.Name
+		if err := cleanupWorkWithLabelsSelectorDelete(c, key,
+			labels.SelectorFromSet(labels.Set{
+				networkingv1alpha1.MultiClusterServiceDerivedWorkNamespace: key.Namespace,
+				networkingv1alpha1.MultiClusterServiceDerivedWorkName:      key.Name,
+				networkingv1alpha1.MultiClusterServiceDerivedWorkKind:      key.Kind,
+			})); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func cleanupWorkWithLabelsSelectorDelete(c client.Client, key keys.FederatedKey, labelsSelector labels.Selector) error {
+	executionSpace := names.GenerateExecutionSpaceName(key.Cluster)
+
+	workList := &workv1alpha1.WorkList{}
+	if err := c.List(context.TODO(), workList, &client.ListOptions{
+		Namespace:     executionSpace,
+		LabelSelector: labelsSelector,
+	}); err != nil {
+		klog.Errorf("Failed to list workList reported by %s(%s) in executionSpace(%s), Error: %v",
+			key.Kind, key.NamespaceKey(), executionSpace, err)
+		return err
+	}
+
+	var errs []error
+	for index, work := range workList.Items {
+		if err := c.Delete(context.TODO(), &workList.Items[index]); err != nil && !apierrors.IsNotFound(err) {
+			klog.Errorf("Failed to delete work(%s/%s), Error: %v", work.Namespace, work.Name, err)
+			errs = append(errs, err)
+		}
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+// needCleanupDerivedWork If ServiceImport or ServiceExport is not found, delete the associated work.
+func needCleanupDerivedWork(obj client.Object) bool {
+	return obj != nil && !obj.GetDeletionTimestamp().IsZero() ||
+		obj != nil && obj.GetDeletionTimestamp().IsZero() &&
+			!controllerutil.ContainsFinalizer(obj, MultiClusterServiceControllerFinalizer)
+}
+
+func getServiceImportOrServiceExport(c client.Client, fedKey keys.FederatedKey) (*unstructured.Unstructured, error) {
+	switch fedKey.Kind {
+	case util.ServiceImportKind:
+		svcImport := &mcsv1alpha1.ServiceImport{}
+		err := c.Get(context.TODO(), types.NamespacedName{Namespace: fedKey.Namespace, Name: fedKey.Name}, svcImport)
+		if err != nil {
+			return nil, err
+		}
+		return helper.ToUnstructured(svcImport)
+
+	case util.ServiceExportKind:
+		svcExport := &mcsv1alpha1.ServiceExport{}
+		err := c.Get(context.TODO(), types.NamespacedName{Namespace: fedKey.Namespace, Name: fedKey.Name}, svcExport)
+		if err != nil {
+			return nil, err
+		}
+		return helper.ToUnstructured(svcExport)
+	}
+	return nil, fmt.Errorf("unsupported kind(%s)", fedKey.Kind)
+}
+
+func handleServiceExportOrServiceExportEvent(c client.Client, record record.EventRecorder, fedKey keys.FederatedKey) (err error) {
```

Comment:
Just one typo: it seems that this function name is `handleServiceImportOrServiceExportEvent`

Reply:
Updated.